ambry-account/src/main/java/com/github/ambry/account/RouterStore.java

```diff
@@ -0,0 +1,331 @@
+    this.accountService = accountService;
+    this.helixStore = helixStore;
+    this.config = config;
+  }
+
+  @Override
+  public void fetchAndUpdateCache(boolean isCalledFromListener) {
+    // when fetching account metadata, we need to fetch the list of blob ids that point to different versions of
+    // account metadata as well the latest version of account metadata.
+    lock.lock();
+    try {
+      long startTimeMs = System.currentTimeMillis();
+      logger.trace("Start reading account metadata blob ids list from path={}", ACCOUNT_METADATA_BLOB_IDS_PATH);
+      ZNRecord zNRecord = helixStore.get(ACCOUNT_METADATA_BLOB_IDS_PATH, null, AccessOption.PERSISTENT);
+      logger.trace("Fetched ZNRecord from path={}, took time={} ms", ACCOUNT_METADATA_BLOB_IDS_PATH, startTimeMs);
+      if (zNRecord == null) {
+        logger.debug("The ZNRecord to read does not exist on path={}", ACCOUNT_METADATA_BLOB_IDS_PATH);
+      } else {
+        List<String> accountBlobIDs = zNRecord.getListField(ACCOUNT_METADATA_BLOB_IDS_LIST_KEY);
+        if (accountBlobIDs == null || accountBlobIDs.size() == 0) {
+          logger.debug("ZNRecord={} to read on path={} does not have a simple list with key={}", zNRecord,
+              ACCOUNT_METADATA_BLOB_IDS_PATH, ACCOUNT_METADATA_BLOB_IDS_LIST_KEY);
+        } else {
+          // parse the json string list and get the blob id with the latest version
+          BlobIDAndVersion blobIDAndVersion = null;
+          for (String accountBlobIDInJson : accountBlobIDs) {
+            BlobIDAndVersion current = BlobIDAndVersion.fromJson(accountBlobIDInJson);
+            if (blobIDAndVersion == null || blobIDAndVersion.version < current.version) {
+              blobIDAndVersion = current;
+            }
+          }
+
+          Map<String, String> accountMap = readAccountMetadataFromBlobID(blobIDAndVersion.blobID);
+          if (accountMap == null) {
+            logger.debug("BlobID={} to read but no account map returned");
+          } else {
+            logger.trace("Start parsing remote account data from blob {} and versioned at {}.", blobIDAndVersion.blobID,
+                blobIDAndVersion.version);
+            AccountInfoMap newAccountInfoMap = new AccountInfoMap(accountService, accountMap);
+            AccountInfoMap oldAccountInfoMap = accountService.updateCache(newAccountInfoMap);
+            accountService.notifyAccountUpdateConsumers(newAccountInfoMap, oldAccountInfoMap, isCalledFromListener);
+          }
+        }
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private Map<String, String> readAccountMetadataFromBlobID(String blobID) {
+    long startTimeMs = System.currentTimeMillis();
+    Future<GetBlobResult> resultF = accountService.router.get().getBlob(blobID, new GetBlobOptionsBuilder().build());
+    try {
+      GetBlobResult result = resultF.get();
+      accountService.accountServiceMetrics.accountFetchFromAmbryTimeInMs.update(
```

Comment:
All the places where you update metrics, please use a saved instance of accountServiceMetrics and don't get it from accountService.